localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -2069,8 +2071,31 @@ def create_function_url_config(
             else api_utils.unqualified_lambda_arn(function_name, account_id, region)
         )
 
-        # create function URL config
-        url_id = api_utils.generate_random_url_id()
+        # The url_id is the subdomain used for the URL we're creating. This
+        # is either created randomly (as in AWS), or can be passed as a tag
+        # to the lambda itself (localstack-only).
+        url_id: str
+        if fn.tags is None:
+            url_id = api_utils.generate_random_url_id()
+
+        else:
+            # Note: currently this only works with $LATEST. However, it would
+            # be nice if we could specify a custom URL for any alias being
+            # passed via the qualifier parameter. Once a strategy for that has
+            # been decided on, it can go here.
+            if normalized_qualifier == "$LATEST" and TAG_KEY_CUSTOM_URL in fn.tags:
+                # Note: I really wanted to add verification here that the
+                # url_id is unique, so we could surface that to the user ASAP.
+                # However, it seems like that information isn't available yet,
+                # since (as far as I can tell) we call
+                # self.router.register_routes() once, in a single shot, for all
+                # of the routes -- and we need to verify that it's unique not
+                # just for this particular lambda function, but for the entire
+                # lambda provider. Therefore... that idea proved non-trivial!
+                url_id = fn.tags[TAG_KEY_CUSTOM_URL]
```

Comment:
I saw a comment about this in the APIGW code, as well as the ``.lower()`` normalization, but I didn't really understand what, exactly, the problem was. AWS tags are always case-sensitive, sure, but domain names are always case-insensitive. Is the function router doing case-sensitive matching or something?

Reply:
Yes, domain name are case insensitive, but always passed as lower cased from our router framework. The issue will be here: https://github.com/localstack/localstack/blob/28d561d2c74ced5322b760ed8d5c80f1f74a88d6/localstack-core/localstack/services/lambda_/urlrouter.py#L65-L70

When we try to look for the function itself.

Reply:
Ah, missed that. Wouldn't it make more sense to apply the ``.lower()`` normalization within the router, on line 69 above? Ie, ``if url_config.url_id.lower() == api_id.lower()``? It seems to me that would be both more robust and more RFC-conformant.

(Side note, arguably the router should be using a precalculated hash lookup instead of nested for loops, but that's well outside the scope of this PR)

Reply:
Yup, both would be working (once again, APIGW was a bit different and I wish we had cast it to lower case when receiving the tag, because you can also use the custom API Id in API Gateway in the `path` part of the URL, which is case-sensitive (and we had to make it case insensitive, thus breaking parity)).

For lambda URL, this does not matter so much, so it is fine to normalize is to lowercase when matching the right function! 

Regarding the hash lookup, we changed it in our ongoing work on APIGW as well just recently 😄 we didn't really have the right tools earlier without complicating things too much.